source/subscription-manager/subscription_manager.h

```diff
@@ -0,0 +1,124 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://aws.amazon.com/freertos
+ *
+ */
+
+/**
+ * @file subscription_manager.h
+ * @brief Functions for managing MQTT subscriptions.
+ */
+#ifndef SUBSCRIPTION_MANAGER_H
+#define SUBSCRIPTION_MANAGER_H
+
+/* Demo config include. */
+#include "demo_config.h"
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+
+/* core MQTT include. */
+#include "core_mqtt.h"
+
+
+/**
+ * @brief Maximum number of subscriptions maintained by the subscription manager
+ * simultaneously in a list.
+ */
+#ifndef SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS
+    #define SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS    10U
+#endif
+
+/**
+ * @brief Callback function called when receiving a publish.
+ *
+ * @param[in] pvIncomingPublishCallbackContext The incoming publish callback context.
+ * @param[in] pxPublishInfo Deserialized publish information.
+ */
+typedef void (* IncomingPubCallback_t )( void * pvIncomingPublishCallbackContext,
+                                         MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief An element in the list of subscriptions maintained by the agent.
+ *
+ * @note The agent allows multiple tasks to subscribe to the same topic.
+ * In this case, another element is added to the subscription list, differing
+ * in the intended publish callback.
+ */
+typedef struct subscriptionElement
+{
+    IncomingPubCallback_t pxIncomingPublishCallback;
+    void * pvIncomingPublishCallbackContext;
+    uint16_t uFilterStringLength;
+    const char * pcSubscriptionFilterString;
+} SubscriptionElement_t;
+
+/**
+ * @brief Add a subscription to the subscription list.
+ *
+ * @note Multiple tasks can be subscribed to the same topic with different
+ * context-callback pairs. However, a single context-callback pair may only be
+ * associated to the same topic filter once.
+ *
+ * @param[in] pxSubscriptionList  The pointer to the subscription list array.
+ * @param[in] pcTopicFilterString Topic filter string of subscription.
+ * @param[in] uTopicFilterLength Length of topic filter string.
+ * @param[in] pxIncomingPublishCallback Callback function for the subscription.
+ * @param[in] pvIncomingPublishCallbackContext Context for the subscription callback.
+ *
+ * @return `true` if subscription added or exists, `false` if insufficient memory.
+ */
+BaseType_t addSubscription( SubscriptionElement_t * pxSubscriptionList,
```

Comment:
Nit: You could save yourself from having to `#include "FreeRTOS.h"` if the return types of these functions were bools or fixed width integers. There's nothing about the subscription manager specific to FreeRTOS.

Reply:
The intention was to avoid the use of bools. However, bool has got introduced with the usage of coreMQTT library. I think it makes sense to use bool so that it is independent of RTOS